akka-remote/src/test/scala/akka/remote/artery/SystemMessageDeliverySpec.scala

```diff
@@ -58,6 +150,87 @@ class SystemMessageDeliverySpec extends AkkaSpec(SystemMessageDeliverySpec.commo
+      ackRecipient.expectMsg(Ack(1L, addressB))
+      ackRecipient.expectMsg(Ack(2L, addressB))
+      // 3 and 4 was dropped
+      ackRecipient.expectMsg(Nack(2L, addressB))
+      sink.expectNoMsg(100.millis) // 3 was dropped
+      replyCallback.invoke(Nack(2L, addressB))
+      // resending 3, 4, 5
+      sink.expectNext("msg-3")
+      ackRecipient.expectMsg(Ack(3L, addressB))
+      sink.expectNext("msg-4")
+      ackRecipient.expectMsg(Ack(4L, addressB))
+      sink.expectNext("msg-5")
+      ackRecipient.expectMsg(Ack(5L, addressB))
+      ackRecipient.expectNoMsg(100.millis)
+      replyCallback.invoke(Ack(5L, addressB))
+      sink.cancel()
+    }
+
+    "be resent when first is lost" in {
+      val ackRecipient = TestProbe()
+      val (sink, replyCallback) =
+        setupManualCallback(ackRecipient.ref, resendInterval = 60.seconds, dropSeqNumbers = Vector(1L), sendCount = 3)
+
+      sink.request(100)
+      ackRecipient.expectMsg(Nack(0L, addressB)) // from receiving 2
+      ackRecipient.expectMsg(Nack(0L, addressB)) // from receiving 3
+      sink.expectNoMsg(100.millis) // 1 was dropped
+      replyCallback.invoke(Nack(0L, addressB))
+      replyCallback.invoke(Nack(0L, addressB))
+      // resending 1, 2, 3
+      sink.expectNext("msg-1")
+      ackRecipient.expectMsg(Ack(1L, addressB))
+      sink.expectNext("msg-2")
+      ackRecipient.expectMsg(Ack(2L, addressB))
+      sink.expectNext("msg-3")
+      ackRecipient.expectMsg(Ack(3L, addressB))
+      replyCallback.invoke(Ack(3L, addressB))
+      sink.cancel()
+    }
+
+    "be resent when last is lost" in {
+      val ackRecipient = TestProbe()
+      val (sink, replyCallback) =
+        setupManualCallback(ackRecipient.ref, resendInterval = 1.second, dropSeqNumbers = Vector(3L), sendCount = 3)
+
+      sink.request(100)
+      sink.expectNext("msg-1")
+      ackRecipient.expectMsg(Ack(1L, addressB))
+      replyCallback.invoke(Ack(1L, addressB))
+      sink.expectNext("msg-2")
+      ackRecipient.expectMsg(Ack(2L, addressB))
+      replyCallback.invoke(Ack(2L, addressB))
+      sink.expectNoMsg(200.millis) // 3 was dropped
+      // resending 3 due to timeout
+      sink.expectNext("msg-3")
+      ackRecipient.expectMsg(Ack(3L, addressB))
+      replyCallback.invoke(Ack(3L, addressB))
+      sink.cancel()
+    }
+
+    "deliver all during random dropping" in {
+      val N = 10000
+      val dropSeqNumbers = (1 to N / 10).map(_ ⇒ ThreadLocalRandom.current().nextInt(1, N).toLong).toVector
+      val sink = setupAutoCallback(resendInterval = 1.second, dropSeqNumbers, sendCount = N)
+      // this test is normally quick (1 second) but since it's random it should not fail due to bad luck
+      sink.within(20.seconds) {
+        sink.request(N / 2)
+        sink.expectNextN(N / 2) should ===((1 to N / 2).map("msg-" + _).toVector)
+        sink.request(N / 2)
+        sink.expectNextN(N / 2) should ===((N / 2 + 1 to N).map("msg-" + _).toVector)
+      }
+    }
+
   }
```

Comment:
Several:
 - this is blocking as it is using probes
 - no fusing, so you cannot benefit from the fuzzer (it should be turned on for stress-testing)
 - since it is probe based, it cannot benefit from composability of streams. For lower level testing of the protocol, probes are fine, but for stress-testing it should be avoided. 

In general, this reads much more like actor testing than stream testing, it just doesn't feel idiomatic streams usage.